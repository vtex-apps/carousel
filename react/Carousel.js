import './global.css'

import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { NoSSR } from 'render'
import { Slider } from 'vtex.store-components'

import Banner from './Banner'

const GLOBAL_PAGES = global.__RUNTIME__ && Object.keys(global.__RUNTIME__.pages)

const defaultBannerProps = index => ({
  image: `https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-0${index +
    1}.png`,
  page: 'store/home',
  description: 'banner',
})

/**
 * Carousel component. Shows a serie of banners.
 */
export default class Carousel extends Component {
  static defaultProps = {
    autoplay: true,
    showArrows: true,
    showDots: true,
    height: 420,
    mobileHeight: 339,
    autoplaySpeed: 5,
    banners: [],
  }

  static uiSchema = {
    banners: {
      items: {
        image: {
          'ui:widget': 'image-uploader',
        },
        typeOfRoute: {
          'ui:widget': 'radio',
          'ui:options': {
            inline: true,
          },
        },
      },
    },
  }

  static propTypes = {
    /** Should change images automatically */
    autoplay: PropTypes.bool.isRequired,
    /** How long it should wait to change the banner in secs */
    autoplaySpeed: PropTypes.number.isRequired,
    /** Max height size of the banners */
    height: PropTypes.number.isRequired,
    /** Max height size of the banners on mobile */
    mobileHeight: PropTypes.number.isRequired,
    /** Set visibility of dots */
    showDots: PropTypes.bool,
    /** Set visibility of arrows */
    showArrows: PropTypes.bool,
    /** Banners that will be displayed by the Carousel */
    banners: PropTypes.arrayOf(
      PropTypes.shape({
        /** The image url of the banner */
        image: PropTypes.string,
        /** The description of the image */
        description: PropTypes.string,
        /** The url where the image is pointing to, in case of external route */
        url: PropTypes.string,
        /** The page where the image is pointing to */
        page: PropTypes.string,
        /** Params of the url */
        params: PropTypes.object,
        /** Indicates if the route is external or not */
        externalRoute: PropTypes.bool,
      })
    ),
  }

  static getSchema = props => {
    const autoplay = props.autoplay || false

    const internalRouteSchema = {
      page: {
        type: 'string',
        enum: GLOBAL_PAGES,
        title: 'editor.carousel.bannerLink.page.title',
        isLayout: false,
      },
      params: {
        type: 'string',
        description: 'editor.carousel.bannerLink.params.description',
        title: 'editor.carousel.bannerLink.params.title',
        isLayout: false,
      },
    }

    const externalRouteSchema = {
      url: {
        type: 'string',
        title: 'editor.carousel.bannerLink.url.title',
        isLayout: false,
      },
    }

    return {
      title: 'editor.carousel.title',
      description: 'editor.carousel.description',
      type: 'object',
      properties: {
        showDots: {
          type: 'boolean',
          title: 'editor.carousel.showDots.title',
          default: true,
          isLayout: true,
        },
        showArrows: {
          type: 'boolean',
          title: 'editor.carousel.showArrows.title',
          default: true,
          isLayout: true,
        },
        height: {
          type: 'number',
          title: 'editor.carousel.height.title',
          default: 420,
          enum: [420, 440],
          isLayout: true,
        },
        mobileHeight: {
          type: 'number',
          title: 'editor.carousel.mobileHeight.title',
          default: 339,
          enum: [339, 159],
          isLayout: true,
        },
        autoplay: {
          type: 'boolean',
          title: 'editor.carousel.autoplay.title',
          default: true,
          isLayout: true,
        },
        autoplaySpeed: autoplay
          ? {
            type: 'number',
            title: 'editor.carousel.autoplaySpeed.title',
            default: 5,
            enum: [4, 5, 6],
            widget: {
              'ui:widget': 'radio',
              'ui:options': {
                inline: true,
              },
            },
            isLayout: true,
          }
          : {},
        banners: {
          type: 'array',
          title: 'Banners',
          minItems: 1,
          items: {
            type: 'object',
            title: 'Banner',
            properties: {
              image: {
                type: 'string',
                title: 'editor.carousel.banner.image.title',
                default: '',
                widget: {
                  'ui:widget': 'image-uploader',
                },
              },
              description: {
                type: 'string',
                title: 'editor.carousel.banner.description.title',
                default: '',
              },
              externalRoute: {
                type: 'boolean',
                title: 'editor.carousel.banner.externalRoute.title',
              },
              ...externalRouteSchema,
              ...internalRouteSchema,
            },
          },
        },
      },
    }
  }

  configureSettings() {
    const { autoplay, autoplaySpeed, showDots, showArrows } = this.props

    return {
      speed: 500,
      infinite: true,
      pauseOnHover: true,
      slidesToShow: 1,
      slidesToScroll: 1,
      adaptiveHeight: false,
      autoplaySpeed: autoplaySpeed * 1000,
      dots: showDots,
      arrows: showArrows,
      autoplay,
    }
  }

  render() {
    const { height, banners, mobileHeight } = this.props
    const settings = this.configureSettings()

    const banner = defaultBannerProps(0)
    const fallback = (
      <div style={{ maxHeight: `${height}px` }} className="overflow-y-hidden">
        <Banner
          height={height}
          image={banner.image}
          description={banner.description}
          mobileHeight={mobileHeight}
          url={banner.url}
          page={banner.page}
          params={banner.params}
          externalRoute={banner.externalRoute}
        />
      </div>
    )
    return (
      <div className="vtex-carousel">
        <NoSSR onSSR={fallback}>
          <Slider sliderSettings={settings}>
            {banners.map(
              (banner, i) =>
                banner &&
                banner.image && (
                  <div key={i} style={{ maxHeight: `${height}px` }}>
                    <Banner
                      height={height}
                      image={banner.image}
                      description={banner.description}
                      mobileHeight={mobileHeight}
                      page={banner.page}
                      params={banner.params}
                      typeOfRoute={banner.typeOfRoute}
                    />
                  </div>
                )
            )}
          </Slider>
        </NoSSR>
      </div>
    )
  }
}
