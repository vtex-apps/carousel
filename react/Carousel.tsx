import classnames from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Dots, Slide, Slider, SliderContainer } from 'vtex.slider'
import { Container } from 'vtex.store-components'
import { IconCaret } from 'vtex.store-icons'

import Banner, { Props as BannerProps } from './Banner'
import styles from './styles.css'

const GLOBAL_PAGES = global.__RUNTIME__ && [
  'Custom',
  ...Object.keys(global.__RUNTIME__.pages),
]

interface Props {
  /** Should change images automatically */
  autoplay?: boolean
  /** How long it should wait to change the banner in secs */
  autoplaySpeed: number | string
  /** Banners that will be displayed by the Carousel */
  banners: BannerProps[]
  /** Max height size of the banners */
  height?: number
  /** Set visibility of arrows */
  showArrows?: boolean
  /** Set visibility of dots */
  showDots?: boolean
}

interface State {
  currentSlide: number
}

interface ArrowProps {
  orientation: string
  onClick: (event: React.MouseEvent<HTMLElement>) => void
}

interface ArrowContainerProps {
  children: React.ReactNode
}

/**
 * Carousel component. Shows a serie of banners.
 */
export default class Carousel extends Component<Props, State> {
  public static defaultProps: Props = {
    autoplay: true,
    autoplaySpeed: 5,
    banners: [],
    height: 420,
    showArrows: true,
    showDots: true,
  }

  public static propTypes = {
    /** Should change images automatically */
    autoplay: PropTypes.bool.isRequired,
    /** How long it should wait to change the banner in secs */
    autoplaySpeed: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
      .isRequired,
    /** Banners that will be displayed by the Carousel */
    banners: PropTypes.arrayOf(
      PropTypes.shape({
        /** The description of the image */
        description: PropTypes.string,
        /** The image url of the banner */
        image: PropTypes.string,
        /** The page where the image is pointing to */
        page: PropTypes.string,
        /** Params of the url */
        params: PropTypes.string,
        /** Indicates if the route is external or internal */
        typeOfRoute: PropTypes.string,
        /** The url where the image is pointing to, in case of external route */
        url: PropTypes.string,
        /** The url where the image is pointing to, in case of internal route (optional) */
        customInternalURL: PropTypes.string,
      })
    ),
    /** Max height size of the banners */
    height: PropTypes.number.isRequired,
    /** Set visibility of arrows */
    showArrows: PropTypes.bool,
    /** Set visibility of dots */
    showDots: PropTypes.bool,
  }

  public static getSchema = () => {
    const internalRouteSchema = {
      page: {
        enum: GLOBAL_PAGES,
        isLayout: false,
        title: 'admin/editor.carousel.bannerLink.page.title',
        type: 'string',
      },
      params: {
        description: 'admin/editor.carousel.bannerLink.params.description',
        isLayout: false,
        title: 'admin/editor.carousel.bannerLink.params.title',
        type: 'string',
      },
      customInternalURL: {
        description:
          'admin/editor.carousel.bannerLink.custominternalurl.description',
        isLayout: false,
        title: 'admin/editor.carousel.bannerLink.custominternalurl.title',
        type: 'string',
      },
    }

    const externalRouteSchema = {
      url: {
        isLayout: false,
        title: 'admin/editor.carousel.bannerLink.url.title',
        type: 'string',
      },
    }

    // tslint:disable:object-literal-sort-keys
    return {
      description: 'admin/editor.carousel.description',
      dependencies: {
        autoplay: {
          oneOf: [
            {
              properties: {
                autoplay: {
                  enum: [true],
                },
                autoplaySpeed: {
                  enum: ['2', '3', '4', '5', '6', '7', '8'],
                  isLayout: false,
                  title: 'admin/editor.carousel.autoplaySpeed.title',
                  type: 'string',
                },
              },
            },
          ],
        },
      },
      properties: {
        banners: {
          items: {
            properties: {
              image: {
                default: '',
                title: 'admin/editor.carousel.banner.image.title',
                type: 'string',
                widget: {
                  'ui:widget': 'image-uploader',
                },
              },
              mobileImage: {
                default: '',
                title: 'admin/editor.carousel.banner.mobileImage.title',
                type: 'string',
                widget: {
                  'ui:widget': 'image-uploader',
                },
              },
              description: {
                default: '',
                title: 'admin/editor.carousel.banner.description.title',
                type: 'string',
              },
              externalRoute: {
                default: false,
                isLayout: false,
                title: 'admin/editor.carousel.banner.externalRoute.title',
                type: 'boolean',
              },
            },
            dependencies: {
              externalRoute: {
                oneOf: [
                  {
                    properties: {
                      externalRoute: {
                        enum: [false],
                      },
                      ...internalRouteSchema,
                    },
                  },
                  {
                    properties: {
                      externalRoute: {
                        enum: [true],
                      },
                      ...externalRouteSchema,
                    },
                  },
                ],
              },
            },
            title: 'admin/editor.carousel.banner.title',
            type: 'object',
          },
          minItems: 1,
          title: 'admin/editor.carousel.banners.title',
          type: 'array',
        },
        height: {
          default: 420,
          enum: [420, 440],
          isLayout: true,
          title: 'admin/editor.carousel.height.title',
          type: 'number',
        },
        showArrows: {
          default: true,
          isLayout: true,
          title: 'admin/editor.carousel.showArrows.title',
          type: 'boolean',
        },
        showDots: {
          default: true,
          isLayout: true,
          title: 'admin/editor.carousel.showDots.title',
          type: 'boolean',
        },
        autoplay: {
          default: true,
          isLayout: false,
          title: 'admin/editor.carousel.autoplay.title',
          type: 'boolean',
        },
      },
      title: 'admin/editor.carousel.title',
      type: 'object',
    }
    // tslint:enable
  }

  public state = {
    currentSlide: 0,
  }

  public perPage = 1

  public handleChangeSlide = (i: number): void => {
    this.setState({ currentSlide: i })
  }

  public handleNextSlide = (): void => {
    this.setState(({ currentSlide }) => {
      const bannersLength: number = this.props.banners.filter(
        banner => banner && (banner.mobileImage || banner.image)
      ).length
      const nextSlide: number =
        ((currentSlide + 1 - this.perPage) % bannersLength) + this.perPage

      return {
        currentSlide: nextSlide,
      }
    })
  }

  public ArrowRender: React.StatelessComponent<ArrowProps> = ({
    orientation,
    onClick,
  }: ArrowProps) => {
    const containerClasses = classnames(styles.arrow, 'pointer z-1 flex', {
      [styles.arrowLeft]: orientation === 'left',
      [styles.arrowRight]: orientation === 'right',
    })
    return (
      <div className={containerClasses} onClick={onClick}>
        <IconCaret orientation={orientation} thin size={20} />
      </div>
    )
  }

  public ArrowContainerRender: React.StatelessComponent<
    ArrowContainerProps
  > = ({ children }: ArrowContainerProps) => {
    const wrapperClasses = classnames(
      styles.arrowsContainerWrapper,
      'w-100 h-100 absolute left-0 top-0 flex justify-center'
    )
    const containerClasses = classnames(
      styles.arrowsContainer,
      'w-100 h-100 mw9 flex-ns justify-between items-center dn-s'
    )

    return (
      <div className={wrapperClasses}>
        <Container className={containerClasses}>{children}</Container>
      </div>
    )
  }

  public render() {
    const { height, showArrows, autoplay, autoplaySpeed, showDots } = this.props
    const { currentSlide } = this.state
    if (!this.props.banners.length) {
      return null
    }

    const banners: BannerProps[] = this.props.banners.filter(
      banner => banner && (banner.mobileImage || banner.image)
    )

    const autoplayInterval = autoplaySpeed
      ? typeof autoplaySpeed === 'string'
        ? parseFloat(autoplaySpeed)
        : autoplaySpeed
      : 0

    return (
      <SliderContainer
        autoplay={autoplay && autoplayInterval > 0}
        autoplayInterval={autoplayInterval * 1000}
        pauseOnHover
        onNextSlide={this.handleNextSlide}
        className={styles.container}
      >
        <Slider
          loop
          classes={{
            root: styles.sliderRoot,
            sliderFrame: styles.sliderFrame,
          }}
          perPage={this.perPage}
          arrowRender={showArrows && this.ArrowRender}
          currentSlide={currentSlide}
          onChangeSlide={this.handleChangeSlide}
          arrowsContainerComponent={showArrows && this.ArrowContainerRender}
          duration={500}
        >
          {banners.map((banner, i) => (
            <Slide
              className={styles.slide}
              key={i}
              style={{ maxHeight: height }}
              sliderTransitionDuration={500}
            >
              <Banner height={height} {...banner} />
            </Slide>
          ))}
        </Slider>
        {showDots && (
          <Dots
            loop
            perPage={this.perPage}
            currentSlide={currentSlide}
            totalSlides={banners.length}
            onChangeSlide={this.handleChangeSlide}
            classes={{
              activeDot: classnames(styles.activeDot, 'bg-emphasis'),
              dot: classnames(styles.dot, 'mh2 mv0 pointer br-100'),
              notActiveDot: classnames(styles.notActiveDot, 'bg-muted-3'),
              root: classnames(styles.containerDots, 'bottom-0 pb4'),
            }}
          />
        )}
      </SliderContainer>
    )
  }
}
