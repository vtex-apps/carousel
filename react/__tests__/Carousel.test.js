/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'

import Carousel from '../Carousel'

describe('Carousel component', () => {
  let wrapper

  const props = {
    height: 440,
    mobileHeight: 339,
    showArrows: true,
    showDots: true,
    autoplay: true,
    autoplaySpeed: 4,
    numberOfBanners: 3,
    banner0: {
      image:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-01.png',
      mobileImage:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-01.png',
      page: '/',
      description: 'banner',
    },
    banner1: {
      image:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-02.png',
      mobileImage:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-02.png',
      page: '/',
      description: 'banner',
    },
    banner2: {
      image:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-03.png',
      mobileImage:
        'https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-03.png',
      page: '/',
      description: 'banner',
    },
  }

  beforeEach(() => {
    wrapper = render(<Carousel {...props} />)
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
  })

  xit('should match snapshot', () => {
    expect(wrapper.container).toMatchSnapshot()
  })

  xit('should render items', () => {
    expect(wrapper.container.querySelectorAll('.vtex-carousel').length).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.vtex-carousel__img-container').length
    ).toBe(3) // set to be infinite
  })
})
