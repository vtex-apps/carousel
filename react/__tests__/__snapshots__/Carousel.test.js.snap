// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component should match snapshot 1`] = `
<div>
  <div
    class="vtex-carousel"
  >
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="img-container"
          >
            <img
              alt="banner"
              class="img-regular w-100"
              src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-01.png"
            />
            <div
              class="img-mobile"
              style="max-height: 339px;"
            >
              <img
                alt="banner"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-01.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="img-container"
          >
            <img
              alt="banner"
              class="img-regular w-100"
              src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-02.png"
            />
            <div
              class="img-mobile"
              style="max-height: 339px;"
            >
              <img
                alt="banner"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-02.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
    <div
      style="max-height: 440px;"
    >
      <div>
        <a>
          <div
            class="img-container"
          >
            <img
              alt="banner"
              class="img-regular w-100"
              src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-03.png"
            />
            <div
              class="img-mobile"
              style="max-height: 339px;"
            >
              <img
                alt="banner"
                class="w-100"
                src="https://raw.githubusercontent.com/vtex-apps/carousel/master/images/banners-mobile-03.png"
              />
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
